### PR TITLE
nginx: optimize locations

### PR DESCRIPTION
**Pull request by Izorkin**
### Small optimize nginx locations.
Processing nginx location before PR:
```
location /api/v1/streaming/ {
```

```
*360 test location: "/"
*360 test location: "sw.js"
*360 test location: "api/v1/streaming/"
*360 test location: ~ "^/(emoji|packs|system/accounts/avatars|system/media_attachments/files)"
*360 using configuration "/api/v1/streaming/"
```

```
location /sw.js {
```
```
*365 test location: "/"
*365 test location: "sw.js"
*365 test location: ~ "^/(emoji|packs|system/accounts/avatars|system/media_attachments/files)"
*365 using configuration "/sw.js"
```
After PR:
```
location ^~ /api/v1/streaming/ {
```
```
*358 test location: "/"
*358 test location: "sw.js"
*358 test location: "api/v1/streaming/"
*358 using configuration "/api/v1/streaming/"
```
```
location = /sw.js 
```
```
*366 test location: "/"
*366 test location: "sw.js"
*366 using configuration "=/sw.js"
```

### Added caching for all static files.
Caching time reduced to 28 days. For files with a unique name and location, an `immutable` cache mode is used.
For other files added cache mode with change detection - `must-revalidate` cache mode. By default, file changes are also checked against with HTTP ETag. Adding caching to `sw.js` file shouldn't cause negative consequences.

### Changed handles 404 page.
Now nginx handles missing static files on its own and throws a 404 error, instead of redirecting to masodon backend.

### Disable regex in locations
More information provided in this report from nginx developer - https://highload.guide/blog/scalable-configuration-nginx.html and https://www.getpagespeed.com/server-setup/nginx-locations-performance-impact-and-optimizations

### CODE REVIEW

**Comment by Gargron** [comment]
> Now nginx handles missing static files on its own and throws a 404 error, instead of redirecting to masodon backend.

This is probably very good for normal deployments but will be an issue for people who use Docker and rely on Rails to serve static assets. I don't know how common that scenario is though.

**Comment by Izorkin** [comment]
> This is probably very good for normal deployments but will be an issue for people who use Docker and rely on Rails to serve static assets. I don't know how common that scenario is though.

Added note.

**Comment by Izorkin** [comment]
Сan also use this variant (instruction is taken from this article https://sleeplessbeastie.eu/2022/05/02/how-to-take-advantage-of-docker-to-install-mastodon/):

> # Extract static files
> Create a temporary volume pointing to `/opt/mastodon/web/public` directory.
> ```
> $ sudo docker volume create --opt type=none --opt device=/opt/mastodon/web/public --opt o=bind temporary_static
> ```
> Copy static files.
> ```
> $ sudo docker run --rm -v "temporary_static:/static" tootsuite/mastodon:v3.5.1 bash -c "cp -r /opt/mastodon/public/* /static/"
> ```
> Remove temporary volume.
> ```
> $ sudo docker volume rm temporary_static
> ```
> That way we can serve these files using nginx.
> ```
> $ ls /opt/mastodon/web/public/
> ```
> ```
> 500.html                    avatars            emoji            inert.css           oops.gif    shortcuts                 web-push-icon_favourite.png
> android-chrome-192x192.png  badge.png          favicon-dev.ico  mask-icon.svg       oops.png    sounds                    web-push-icon_reblog.png
> apple-touch-icon.png        browserconfig.xml  favicon.ico      mstile-150x150.png  packs       sw.js
> assets                      embed.js           headers          ocr                 robots.txt  web-push-icon_expand.png
> ```
> Delete and copy static files after every mastodon container upgrade.

**Comment by Izorkin** [comment]
Updated and rebased PR.

File: dist/nginx.conf [hunk 1]
```diff
@@ -56,58 +56,104 @@ server {
...
+  location = sw.js {
+    add_header Cache-Control "public, max-age=604800, must-revalidate";
     add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
-    try_files $uri @proxy;
+    try_files $uri =404;
   }
 
-  location /sw.js {
-    add_header Cache-Control "public, max-age=0";
+  location ~ ^/assets/) {
```
Comment (Gargron) [inline dist/nginx.conf:67]: pcre_compile() failed: unmatched parentheses in "^/assets/

**Review by Gargron — CHANGES_REQUESTED** [review]

File: dist/nginx.conf (hunk 1 above)
Comment (Izorkin) [reply dist/nginx.conf:67]: Are you checking with nginx? Which version?

Comment (Gargron) [reply dist/nginx.conf:67]: There is an unmatched `)`. Is that not a mistake?

Comment (Izorkin) [reply dist/nginx.conf:67]: Ups...

Comment (Izorkin) [reply dist/nginx.conf]: Fixed!

**Review by Gargron — APPROVED** [review]

File: dist/nginx.conf [hunk 2]
```diff
@@ -56,58 +56,104 @@ server {
...
+    add_header Cache-Control "public, max-age=2419200, must-revalidate";
+    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
+    try_files $uri =404;
+  }
+
+  location ~ ^/shortcuts/ {
+    add_header Cache-Control "public, max-age=2419200, must-revalidate";
+    add_header Strict-Transport-Security "max-age=63072000; includeSubDomains";
+    try_files $uri =404;
+  }
```
Comment (ClearlyClaire) [inline dist/nginx.conf:101]: What's this?

Comment (ClearlyClaire) [reply dist/nginx.conf:101]: Ah, this was indeed added in #15234 (included in v3.3.0) and removed in #18592 (included in v4.0.0)!

**Pull request merged**